server/src/metrics/client_impact_metrics.rs

```diff
@@ -0,0 +1,298 @@
+use crate::metrics::client_metrics::MetricsCache;
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::hash::{Hash};
+use unleash_types::client_metrics::{ImpactMetric, ImpactMetricEnv};
+use unleash_types::Merge;
+use utoipa::ToSchema;
+
+#[derive(Debug, Clone, Eq, Deserialize, Serialize, ToSchema, Hash, PartialEq)]
+pub struct ImpactMetricsKey {
+    pub app_name: String,
+    pub environment: String,
+}
+
+impl From<ImpactMetricEnv> for ImpactMetricsKey {
+    fn from(value: ImpactMetricEnv) -> Self {
+        Self {
+            app_name: value.app_name,
+            environment: value.environment,
+        }
+    }
+}
+
+pub fn convert_to_impact_metrics_env(metrics: Vec<ImpactMetric>, app_name: String, environment: String) -> Vec<ImpactMetricEnv> {
+    metrics.into_iter()
+        .map(|metric| ImpactMetricEnv::new(metric, app_name.clone(), environment.clone()))
+        .collect()
+}
+
+fn group_by_key(impact_metrics: Vec<ImpactMetricEnv>) -> HashMap<ImpactMetricsKey, Vec<ImpactMetricEnv>> {
+    impact_metrics
+        .into_iter()
+        .chunk_by(|m| ImpactMetricsKey::from(m.clone()))
+        .into_iter()
+        .map(|(k, group)| (k, group.collect()))
+        .collect()
+}
+
+fn index_by_name(metrics: Vec<ImpactMetricEnv>) -> HashMap<String, ImpactMetricEnv> {
+    metrics
+        .into_iter()
+        .map(|metric| (metric.impact_metric.name.clone(), metric))
+        .collect()
+}
+
+fn reduce_metrics_samples(metric: ImpactMetricEnv) -> ImpactMetricEnv {
+    let empty_metric = ImpactMetricEnv::new(
+        ImpactMetric {
+            name: metric.impact_metric.name.clone(),
+            help: metric.impact_metric.help.clone(),
+            r#type: metric.impact_metric.r#type.clone(),
+            samples: vec![],
+        },
+        metric.app_name.clone(),
+        metric.environment.clone(),
+    );
+
+    empty_metric.merge(metric)
+}
+
+pub fn merge_impact_metrics(metrics: Vec<ImpactMetricEnv>) -> Vec<ImpactMetricEnv> {
+    if metrics.is_empty() {
+        return Vec::new();
+    }
+
+    let mut merged_metrics: HashMap<String, ImpactMetricEnv> = HashMap::with_capacity(metrics.len());
+
+    for metric in metrics {
+        let metric_name = metric.impact_metric.name.clone();
+
+        if let Some(existing_metric) = merged_metrics.get_mut(&metric_name) {
+            let merged = existing_metric.clone().merge(metric);
```

Comment:
I don't like line 73 and 74 but don't know how to simplify it with the current merge trait semantics